### PR TITLE
Fix product_uuid in entrypoint

### PR DESCRIPTION
**Pull request by fabiand**
Addresses #1027

Signed-off-by: Fabian Deutsch <fabiand@fedoraproject.org>

### CODE REVIEW

File: images/base/files/usr/local/bin/entrypoint [hunk 1]
```diff
@@ -88,6 +88,23 @@ fix_product_name() {
   fi
 }
 
+fix_product_uuid() {
+  # The system UUID is usually read from DMI via sysfs, the problem is that
+  # in the kind case this means that all (container) nodes shar ethe same
```
[the remark this answers is not in the document]
Comment (fabiand) [reply images/base/files/usr/local/bin/entrypoint]: Thanks, addressed.

File: images/base/files/usr/local/bin/entrypoint [hunk 2]
```diff
@@ -88,6 +88,26 @@ fix_product_name() {
...
+  # The system UUID is usually read from DMI via sysfs, the problem is that
+  # in the kind case this means that all (container) nodes share the same
+  # system/product uuid, as they share the same DMI.
+  # Note: The UUID is read from DMI, this tool is overwriting the sysfs files
+  # which should fix the attached issue, but this workaround does not address
+  # the issue if a tool is reading directly from DMI.
+  # https://github.com/kubernetes-sigs/kind/issues/1027
+  if [[ -f /sys/class/dmi/id/product_uuid ]]; then
+    echo 'INFO: faking /sys/class/dmi/id/product_uuid to be random'
+    cat /proc/sys/kernel/random/uuid > /kind/product_uuid
```
Comment (aojea) [inline images/base/files/usr/local/bin/entrypoint]: @BenTheElder is this compatible with all OSs?

**Comment by aojea** [comment]
/ok-to-test
cool, thanks

**Comment by fabiand** [comment]
I've at least verified this inside the unpatched node image - it worked.

>

**Comment by fabiand** [comment]
/retest

but probably not whitelisted I am

File: images/base/files/usr/local/bin/entrypoint (hunk 2 above)
Comment (BenTheElder) [reply images/base/files/usr/local/bin/entrypoint]: i'm still looking into that aspect, it looks like this is standard.

File: images/base/files/usr/local/bin/entrypoint [hunk 3]
```diff
@@ -88,6 +88,25 @@ fix_product_name() {
...
 
+fix_product_uuid() {
+  # The system UUID is usually read from DMI via sysfs, the problem is that
+  # in the kind case this means that all (container) nodes share the same
+  # system/product uuid, as they share the same DMI.
+  # Note: The UUID is read from DMI, this tool is overwriting the sysfs files
+  # which should fix the attached issue, but this workaround does not address
+  # the issue if a tool is reading directly from DMI.
+  # https://github.com/kubernetes-sigs/kind/issues/1027
+  [[ ! -f /kind/product_uuid ]] && cat /proc/sys/kernel/random/uuid > /kind/product_uuid
```
Comment (BenTheElder) [inline images/base/files/usr/local/bin/entrypoint]: minor tweak: 
- create IFF the file doesn't already exist (for eventual restart node support in kind...)
- decouple the two IFs below by bind mounting both from this faked file

File: pkg/build/node/node.go [hunk 4]
```diff
@@ -43,7 +43,7 @@ import (
 const DefaultImage = "kindest/node:latest"
 
 // DefaultBaseImage is the default base image used
-const DefaultBaseImage = "kindest/base:v20191025-0fae556c@sha256:6c1479d05220704be6c0a69692c1525338381fd8e27093f1268fe747b536360c"
+const DefaultBaseImage = "kindest/base:v20191030-ae10d883@sha256:bccb2ad3686a11ea920b5eaf701b252166affa3c3362e38c549374f1e467ae80"
```
Comment (BenTheElder) [inline pkg/build/node/node.go]: pushed images based on this change

File: images/base/files/usr/local/bin/entrypoint [hunk 5]
```diff
@@ -88,6 +88,25 @@ fix_product_name() {
...
+  if [[ -f /sys/class/dmi/id/product_uuid ]]; then
+    echo 'INFO: faking /sys/class/dmi/id/product_uuid to be random'
+    mount -o ro,bind /kind/product_uuid /sys/class/dmi/id/product_uuid
+  fi
+  if [[ -f /sys/devices/virtual/dmi/id/product_uuid ]]; then
```
Comment (BenTheElder) [inline images/base/files/usr/local/bin/entrypoint]: TODO: determine if we should just always fake these unconditionally for consistency :thinking: 

this should be fine for now

**Comment by BenTheElder** [comment]
thanks for this!

also noting 
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#verify-the-mac-address-and-product-uuid-are-unique-for-every-node actually has this documented upstream :upside_down_face:

**Comment by fabiand** [comment]
> also noting
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#verify-the-mac-address-and-product-uuid-are-unique-for-every-node actually has this documented upstream upside_down_face

Funny.

Thanks for taking care of the additional cases!

I'll test the images

**Comment by BenTheElder** [comment]
works on my machine :tm: :+1: 

we should be able to see the "node logs" from the CI results as well :smile:

**Comment by fabiand** [comment]
Output on my node:
```
$ kubectl get nodes -o yaml | grep -i uuid
      systemUUID: 06cda4eb-cf22-4e79-9dd4-75e2567f14b3
      systemUUID: f1b6214e-f5c6-4663-b859-99b09de62a98
      systemUUID: 34792f2b-765c-4c10-9ec5-1fc7dd9994d5
```

**Pull request merged**